File: src/core/runtime.ts
import { createHmac, timingSafeEqual } from "node:crypto";
import { createActionEnvelope, InMemoryActionBus, type RuntimeActionEnvelope } from "./actions.js";
import { type AuditLogEntry, type AuditStatus } from "./audit.js";

export type ServerStatus = "online" | "offline";
export type RuntimeInstanceStatus = "online" | "degraded" | "offline";
export type RuntimeHealthStatus = "online" | "degraded" | "offline";

export interface ServerRecord {
  id: string;
  name: string;
  environment: string;
  publicKey: string;
  status: ServerStatus;
  lastHeartbeatAt: Date;
}

export interface RuntimeInstanceRecord {
  id: string;
  serverId: string;
  resourceVersion: string;
  fxserverBuild: string;
  gameBuild: string;
  status: RuntimeInstanceStatus;
  startedAt: Date;
  lastSeenAt: Date;
}

export interface RuntimeConfigRecord {
  id: string;
  serverId: string;
  namespace: string;
  key: string;
  value: unknown;
  version: number;
  updatedAt: Date;
}

export interface RuntimeConfigAckRecord {
  serverId: string;
  namespace: string;
  key: string;
  version: number;
  acknowledgedAt: Date;
}

export interface RuntimeHealthRecord {
  serverId: string;
  serverName: string;
  environment: string;
  status: RuntimeHealthStatus;
  reason: string;
  resourceVersion?: string;
  fxserverBuild?: string;
  gameBuild?: string;
  lastHeartbeatAt: Date;
  lastSeenAt?: Date;
}

export interface RegisterServerInput {
  id: string;
  name: string;
  environment: string;
  publicKey: string;
}

export interface HeartbeatInput {
  serverId: string;
  resourceVersion: string;
  fxserverBuild: string;
  gameBuild: string;
  nonce?: string;
  signature?: string;
}

export type HeartbeatSigningInput = Omit<HeartbeatInput, "signature">;

export interface SetRuntimeConfigInput {
  serverId: string;
  namespace: string;
  key: string;
  value: unknown;
}

export interface AckConfigVersionInput {
  serverId: string;
  namespace: string;
  key: string;
  version: number;
}

export interface SubmitActionInput {
  serverId: string;
  actorId: string;
  actionType: string;
  payload: unknown;
  nonce: string;
  idempotencyKey: string;
  signature?: string;
}

export interface CompleteActionInput {
  actionId: string;
  status: "completed" | "failed";
  pluginId?: string;
  permissionKey?: string;
  targetType?: string;
  targetId?: string;
  before?: unknown;
  after?: unknown;
}

export interface WriteAuditLogInput {
  serverId: string;
  actorId: string;
  pluginId?: string;
  actionType: string;
  permissionKey?: string;
  targetType?: string;
  targetId?: string;
  before?: unknown;
  after?: unknown;
  status: AuditLogEntry["status"];
}

export interface RuntimeControlPlaneOptions {
  now?: () => Date;
  idFactory?: () => string;
  actionSignatureVerifier?: (envelope: RuntimeActionEnvelope) => boolean;
  heartbeatSignatureVerifier?: (server: ServerRecord, heartbeat: HeartbeatInput) => boolean;
}

export function signHeartbeat(heartbeat: HeartbeatSigningInput, secret: string): string {
  return createHmac("sha256", secret).update(heartbeatSigningPayload(heartbeat)).digest("hex");
}

export function verifyHeartbeatSignature(
  server: ServerRecord,
  heartbeat: HeartbeatInput,
  secret: string
): boolean {
  if (!heartbeat.signature || !heartbeat.nonce) {
    return false;
  }

  const expected = signHeartbeat(heartbeat, secret);
  const actual = Buffer.from(heartbeat.signature, "hex");
  const expectedBuffer = Buffer.from(expected, "hex");
  return server.id === heartbeat.serverId &&
    actual.length === expectedBuffer.length &&
    timingSafeEqual(actual, expectedBuffer);
}

export class RuntimeControlPlane {
  private readonly serversById = new Map<string, ServerRecord>();
  private readonly instancesByServerId = new Map<string, RuntimeInstanceRecord>();
  private readonly configByKey = new Map<string, RuntimeConfigRecord>();
  private readonly configAcksByKey = new Map<string, RuntimeConfigAckRecord>();
  private readonly actions: InMemoryActionBus;
  private readonly auditLogs: AuditLogEntry[] = [];
  private readonly heartbeatNoncesByServerId = new Map<string, Set<string>>();
  private readonly now: () => Date;
  private readonly idFactory: () => string;
  private readonly heartbeatSignatureVerifier?: (server: ServerRecord, heartbeat: HeartbeatInput) => boolean;

  public constructor(options: RuntimeControlPlaneOptions = {}) {
    this.now = options.now ?? (() => new Date());
    this.idFactory = options.idFactory ?? (() => crypto.randomUUID());
    this.heartbeatSignatureVerifier = options.heartbeatSignatureVerifier;
    this.actions = new InMemoryActionBus({
      verifySignature: options.actionSignatureVerifier
    });
  }

  public registerServer(input: RegisterServerInput): ServerRecord {
    const timestamp = this.now();
    const existing = this.serversById.get(input.id);
    const server: ServerRecord = {
      id: input.id,
      name: input.name,
      environment: input.environment,
      publicKey: input.publicKey,
      status: "online",
      lastHeartbeatAt: timestamp
    };

    this.serversById.set(input.id, {
      ...existing,
      ...server
    });

    return { ...server };
  }

  public heartbeat(input: HeartbeatInput): RuntimeInstanceRecord {
    const server = this.requireServer(input.serverId);
    this.verifyHeartbeat(server, input);

    const timestamp = this.now();
    const existing = this.instancesByServerId.get(input.serverId);
    const instance: RuntimeInstanceRecord = {
      id: existing?.id ?? this.idFactory(),
      serverId: input.serverId,
      resourceVersion: input.resourceVersion,
      fxserverBuild: input.fxserverBuild,
      gameBuild: input.gameBuild,
      status: "online",
      startedAt: existing?.startedAt ?? timestamp,
      lastSeenAt: timestamp
    };

    this.instancesByServerId.set(input.serverId, instance);
    this.serversById.set(input.serverId, {
      ...server,
      status: "online",
      lastHeartbeatAt: timestamp
    });

    return { ...instance };
  }

  private verifyHeartbeat(server: ServerRecord, input: HeartbeatInput): void {
    if (!this.heartbeatSignatureVerifier) {
      return;
    }

    if (!input.nonce || !input.signature) {
      throw new Error("Signed heartbeat requires nonce and signature");
    }

    const usedNonces = this.heartbeatNoncesByServerId.get(server.id) ?? new Set<string>();
    if (usedNonces.has(input.nonce)) {
      throw new Error("Heartbeat nonce replay");
    }

    if (!this.heartbeatSignatureVerifier(server, input)) {
      throw new Error("Invalid heartbeat signature");
    }

    usedNonces.add(input.nonce);
    this.heartbeatNoncesByServerId.set(server.id, usedNonces);
  }

  public setRuntimeConfig(input: SetRuntimeConfigInput): RuntimeConfigRecord {
    this.requireServer(input.serverId);

    const key = this.configKey(input.serverId, input.namespace, input.key);
    const existing = this.configByKey.get(key);
    const record: RuntimeConfigRecord = {
      id: existing?.id ?? this.idFactory(),
      serverId: input.serverId,
      namespace: input.namespace,
      key: input.key,
      value: structuredClone(input.value),
      version: (existing?.version ?? 0) + 1,
      updatedAt: this.now()
    };

    this.configByKey.set(key, record);
    return this.cloneConfig(record);
  }

  public getRuntimeConfig(
    serverId: string,
    namespace: string,
    key: string
  ): RuntimeConfigRecord | undefined {
    const record = this.configByKey.get(this.configKey(serverId, namespace, key));
    return record ? this.cloneConfig(record) : undefined;
  }

  public getConfigSnapshot(serverId: string): RuntimeConfigRecord[] {
    return [...this.configByKey.values()]
      .filter((record) => record.serverId === serverId)
      .sort((a, b) => a.namespace.localeCompare(b.namespace) || a.key.localeCompare(b.key))
      .map((record) => this.cloneConfig(record));
  }

  public getHealth(serverId: string, heartbeatTimeoutMs = 30_000): RuntimeHealthRecord {
    const server = this.requireServer(serverId);
    const instance = this.instancesByServerId.get(serverId);
    const now = this.now();
    const heartbeatAgeMs = now.getTime() - server.lastHeartbeatAt.getTime();

    if (!instance) {
      return {
        serverId: server.id,
        serverName: server.name,
        environment: server.environment,
        status: "degraded",
        reason: "no runtime heartbeat",
        lastHeartbeatAt: new Date(server.lastHeartbeatAt.getTime())
      };
    }

    const health: RuntimeHealthRecord = {
      serverId: server.id,
      serverName: server.name,
      environment: server.environment,
      status: instance.status,
      reason: "runtime heartbeat current",
      resourceVersion: instance.resourceVersion,
      fxserverBuild: instance.fxserverBuild,
      gameBuild: instance.gameBuild,
      lastHeartbeatAt: new Date(server.lastHeartbeatAt.getTime()),
      lastSeenAt: new Date(instance.lastSeenAt.getTime())
    };

    if (server.status === "offline" || instance.status === "offline") {
      return {
        ...health,
        status: "offline",
        reason: "runtime marked offline"
      };
    }
    if (heartbeatAgeMs > heartbeatTimeoutMs) {
      return {
        ...health,
        status: "offline",
        reason: "runtime heartbeat stale"
      };
    }
    if (instance.status === "degraded") {
      return {
        ...health,
        status: "degraded",
        reason: "runtime marked degraded"
      };
    }

    return health;
  }

  public ackConfigVersion(input: AckConfigVersionInput): RuntimeConfigAckRecord {
    this.requireServer(input.serverId);

    const configKey = this.configKey(input.serverId, input.namespace, input.key);
    const config = this.configByKey.get(configKey);
    if (!config) {
      throw new Error(`Unknown config: ${configKey}`);
    }
    if (config.version !== input.version) {
      throw new Error(
        `Cannot acknowledge config version ${input.version}; current version is ${config.version}`
      );
    }

    const ack: RuntimeConfigAckRecord = {
      serverId: input.serverId,
      namespace: input.namespace,
      key: input.key,
      version: input.version,
      acknowledgedAt: this.now()
    };
    this.configAcksByKey.set(configKey, ack);
    return this.cloneConfigAck(ack);
  }

  public getConfigAck(
    serverId: string,
    namespace: string,
    key: string
  ): RuntimeConfigAckRecord | undefined {
    const record = this.configAcksByKey.get(this.configKey(serverId, namespace, key));
    return record ? this.cloneConfigAck(record) : undefined;
  }

  public submitAction(input: SubmitActionInput): RuntimeActionEnvelope {
    this.requireServer(input.serverId);

    return this.actions.submit(
      createActionEnvelope({
        id: this.idFactory(),
        serverId: input.serverId,
        actorId: input.actorId,
        actionType: input.actionType,
        payload: input.payload,
        nonce: input.nonce,
        idempotencyKey: input.idempotencyKey,
        signature: input.signature,
        createdAt: this.now()
      })
    );
  }

  public getAction(actionId: string): RuntimeActionEnvelope | undefined {
    const action = this.actions.getById(actionId);
    return action ? { ...action } : undefined;
  }

  public completeAction(input: CompleteActionInput): RuntimeActionEnvelope {
    const action = this.actions.getById(input.actionId);
    if (!action) {
      throw new Error(`Unknown action: ${input.actionId}`);
    }

    const completedAt = this.now();
    const completed = this.actions.complete(action.id, input.status, completedAt);

    this.auditLogs.push({
      id: this.idFactory(),
      serverId: action.serverId,
      actorId: action.actorId,
      pluginId: input.pluginId,
      actionType: action.actionType,
      permissionKey: input.permissionKey,
      targetType: input.targetType,
      targetId: input.targetId,
      before: structuredClone(input.before),
      after: structuredClone(input.after),
      status: input.status === "completed" ? "succeeded" : "failed",
      createdAt: completedAt
    });

    return completed;
  }

  public writeAuditLog(input: WriteAuditLogInput): AuditLogEntry {
    this.requireServer(input.serverId);
    validateAuditStatus(input.status);

    const entry: AuditLogEntry = {
      id: this.idFactory(),
      serverId: input.serverId,
      actorId: input.actorId,
      pluginId: input.pluginId,
      actionType: input.actionType,
      permissionKey: input.permissionKey,
      targetType: input.targetType,
      targetId: input.targetId,
      before: structuredClone(input.before),
      after: structuredClone(input.after),
      status: input.status,
      createdAt: this.now()
    };
    this.auditLogs.push(entry);
    return {
      ...entry,
      before: structuredClone(entry.before),
      after: structuredClone(entry.after)
    };
  }

  public getAuditLogs(serverId: string): AuditLogEntry[] {
    return this.auditLogs
      .filter((entry) => entry.serverId === serverId)
      .map((entry) => ({
        ...entry,
        before: structuredClone(entry.before),
        after: structuredClone(entry.after)
      }));
  }

  private requireServer(serverId: string): ServerRecord {
    const server = this.serversById.get(serverId);
    if (!server) {
      throw new Error(`Unknown server: ${serverId}`);
    }

    return server;
  }

  private configKey(serverId: string, namespace: string, key: string): string {
    return `${serverId}:${namespace}:${key}`;
  }

  private cloneConfig(record: RuntimeConfigRecord): RuntimeConfigRecord {
    return {
      ...record,
      value: structuredClone(record.value)
    };
  }

  private cloneConfigAck(record: RuntimeConfigAckRecord): RuntimeConfigAckRecord {
    return {
      ...record,
      acknowledgedAt: new Date(record.acknowledgedAt.getTime())
    };
  }
}

function validateAuditStatus(status: AuditStatus): void {
  if (status !== "succeeded" && status !== "failed" && status !== "denied") {
    throw new Error(`Invalid audit status: ${status}`);
  }
}

function heartbeatSigningPayload(heartbeat: HeartbeatSigningInput): string {
  return [
    heartbeat.serverId,
    heartbeat.resourceVersion,
    heartbeat.fxserverBuild,
    heartbeat.gameBuild,
    heartbeat.nonce
  ].join("\n");
}

File: spacetimedb/src/runtime.rs
#[reducer]
pub fn register_server(
    ctx: &ReducerContext,
    id: String,
    name: String,
    environment: String,
    public_key: String,
) -> Result<(), String> {
    validate_server_registration_identity(&id, &name, &environment, &public_key)?;
    let now = ctx.timestamp;
    let server = Server {
        id: id.clone(),
        name,
        environment,
        public_key,
        status: "online".to_string(),
        last_heartbeat_at: now,
    };

    if ctx.db.servers().id().find(id).is_some() {
        ctx.db.servers().id().update(server);
        return Ok(());
    }

    ctx.db.servers().insert(server);
    Ok(())
}

fn validate_server_registration_identity(
    id: &str,
    name: &str,
    environment: &str,
    public_key: &str,
) -> Result<(), String> {
    if id.trim().is_empty() {
        return Err("server id is required".to_string());
    }
    if name.trim().is_empty() {
        return Err("server name is required".to_string());
    }
    if environment.trim().is_empty() {
        return Err("server environment is required".to_string());
    }
    if public_key.trim().is_empty() {
        return Err("server public key is required".to_string());
    }
    Ok(())
}

#[reducer]
pub fn heartbeat(
    ctx: &ReducerContext,
    id: String,
    server_id: String,
    resource_version: String,
    fxserver_build: String,
    game_build: String,
    nonce: String,
    signature: String,
) -> Result<(), String> {
    validate_heartbeat_identity(&id, &server_id, &resource_version, &fxserver_build, &game_build, &nonce, &signature)?;
    ensure_heartbeat_replay_safe(ctx, &server_id, &nonce, &signature)?;
    record_server_heartbeat(ctx, &server_id)?;

    let now = ctx.timestamp;
    ctx.db.runtime_instances().insert(RuntimeInstance {
        id,
        server_id: server_id.clone(),
        resource_version,
        fxserver_build,
        game_build,
        status: "online".to_string(),
        started_at: now,
        last_seen_at: now,
    });
    ctx.db.heartbeat_nonces().insert(HeartbeatNonce {
        id: heartbeat_nonce_id(&server_id, &nonce),
        server_id,
        nonce,
        signature,
        seen_at: now,
    });
    Ok(())
}

fn validate_heartbeat_identity(
    id: &str,
    server_id: &str,
    resource_version: &str,
    fxserver_build: &str,
    game_build: &str,
    nonce: &str,
    signature: &str,
) -> Result<(), String> {
    if id.trim().is_empty() {
        return Err("heartbeat runtime instance id is required".to_string());
    }
    if server_id.trim().is_empty() {
        return Err("heartbeat server id is required".to_string());
    }
    if resource_version.trim().is_empty() {
        return Err("heartbeat resource version is required".to_string());
    }
    if fxserver_build.trim().is_empty() {
        return Err("heartbeat fxserver build is required".to_string());
    }
    if game_build.trim().is_empty() {
        return Err("heartbeat game build is required".to_string());
    }
    if nonce.trim().is_empty() {
        return Err("heartbeat nonce is required".to_string());
    }
    if signature.trim().is_empty() {
        return Err("heartbeat signature is required".to_string());
    }
    Ok(())
}

fn ensure_heartbeat_replay_safe(
    ctx: &ReducerContext,
    server_id: &str,
    nonce: &str,
    signature: &str,
) -> Result<(), String> {
    if signature.is_empty() || nonce.is_empty() {
        return Err("heartbeat signature is required".to_string());
    }

    if ctx
        .db
        .heartbeat_nonces()
        .iter()
        .any(|heartbeat| heartbeat.server_id == server_id && heartbeat.nonce == nonce)
    {
        return Err("heartbeat nonce replay".to_string());
    }

    Ok(())
}

fn record_server_heartbeat(ctx: &ReducerContext, server_id: &str) -> Result<(), String> {
    let Some(mut server) = ctx.db.servers().id().find(server_id.to_string()) else {
        return Err("unknown server".to_string());
    };

    server.last_heartbeat_at = ctx.timestamp;
    server.status = "online".to_string();
    ctx.db.servers().id().update(server);
    Ok(())
}

fn heartbeat_nonce_id(server_id: &str, nonce: &str) -> String {
    format!("{server_id}:{nonce}")
}

#[reducer]
pub fn submit_action(
    ctx: &ReducerContext,
    id: String,
    server_id: String,
    actor_id: String,
    action_type: String,
    payload_hash: String,
    signature: String,
    nonce: String,
    idempotency_key: String,
) -> Result<(), String> {
    validate_runtime_action_envelope(&id, &server_id, &actor_id, &action_type, &payload_hash, &signature, &nonce, &idempotency_key)?;
    if !server_exists(ctx, &server_id) {
        return Err("server must exist before action writes".to_string());
    }
    if ensure_runtime_action_replay_safe(
        ctx,
        &server_id,
        &actor_id,
        &action_type,
        &payload_hash,
        &signature,
        &nonce,
        &idempotency_key,
    )? {
        return Ok(());
    }

    ctx.db.runtime_actions().insert(RuntimeAction {
        id,
        server_id,
        actor_id,
        action_type,
        payload_hash,
        signature,
        nonce,
        idempotency_key,
        status: "submitted".to_string(),
        created_at: ctx.timestamp,
        completed_at: None,
    });
    Ok(())
}

fn validate_runtime_action_envelope(
    id: &str,
    server_id: &str,
    actor_id: &str,
    action_type: &str,
    payload_hash: &str,
    signature: &str,
    nonce: &str,
    idempotency_key: &str,
) -> Result<(), String> {
    if id.trim().is_empty() {
        return Err("runtime action id is required".to_string());
    }
    if server_id.trim().is_empty() {
        return Err("runtime action server id is required".to_string());
    }
    if actor_id.trim().is_empty() {
        return Err("runtime action actor id is required".to_string());
    }
    if action_type.trim().is_empty() {
        return Err("runtime action type is required".to_string());
    }
    if payload_hash.trim().is_empty() {
        return Err("runtime action payload hash is required".to_string());
    }
    if signature.trim().is_empty() {
        return Err("runtime action signature is required".to_string());
    }
    if nonce.trim().is_empty() {
        return Err("runtime action nonce is required".to_string());
    }
    if idempotency_key.trim().is_empty() {
        return Err("runtime action idempotency key is required".to_string());
    }
    Ok(())
}

fn ensure_runtime_action_replay_safe(
    ctx: &ReducerContext,
    server_id: &str,
    actor_id: &str,
    action_type: &str,
    payload_hash: &str,
    signature: &str,
    nonce: &str,
    idempotency_key: &str,
) -> Result<bool, String> {
    if signature.is_empty() {
        return Err("action signature is required".to_string());
    }

    for action in ctx.db.runtime_actions().iter() {
        if action.server_id == server_id && action.idempotency_key == idempotency_key {
            if action.actor_id == actor_id
                && action.action_type == action_type
                && action.payload_hash == payload_hash
            {
                return Ok(true);
            }

            return Err("action idempotency conflict".to_string());
        }

        if action.server_id == server_id && action.nonce == nonce {
            return Err("nonce replay".to_string());
        }
    }

    Ok(false)
}

#[reducer]
pub fn complete_action(ctx: &ReducerContext, action_id: String, status: String) -> Result<(), String> {
    validate_runtime_action_completion_identity(&action_id)?;
    if status != "completed" && status != "failed" {
        return Err("status must be completed or failed".to_string());
    }

    if let Some(mut action) = ctx.db.runtime_actions().id().find(action_id) {
        if action.completed_at.is_some() {
            return Err("action is already completed".to_string());
        }
        action.status = status;
        action.completed_at = Some(ctx.timestamp);
        ctx.db.runtime_actions().id().update(action);
        return Ok(());
    }

    Err("unknown action".to_string())
}

fn validate_runtime_action_completion_identity(action_id: &str) -> Result<(), String> {
    if action_id.trim().is_empty() {
        return Err("runtime action completion id is required".to_string());
    }
    Ok(())
}

#[reducer]
pub fn set_runtime_config(
    ctx: &ReducerContext,
    id: String,
    server_id: String,
    namespace: String,
    key: String,
    value_json: String,
    version: u64,
) -> Result<(), String> {
    validate_runtime_config_identity(&id, &server_id, &namespace, &key, &value_json)?;
    if !server_exists(ctx, &server_id) {
        return Err("server must exist before config writes".to_string());
    }

    ctx.db.runtime_config().insert(RuntimeConfig {
        id,
        server_id,
        namespace,
        key,
        value_json,
        version,
        updated_at: ctx.timestamp,
    });
    Ok(())
}

fn validate_runtime_config_identity(
    id: &str,
    server_id: &str,
    namespace: &str,
    key: &str,
    value_json: &str,
) -> Result<(), String> {
    if id.trim().is_empty() {
        return Err("runtime config id is required".to_string());
    }
    if server_id.trim().is_empty() {
        return Err("runtime config server id is required".to_string());
    }
    if namespace.trim().is_empty() {
        return Err("runtime config namespace is required".to_string());
    }
    if key.trim().is_empty() {
        return Err("runtime config key is required".to_string());
    }
    if value_json.trim().is_empty() {
        return Err("runtime config value json is required".to_string());
    }
    Ok(())
}

#[reducer]
pub fn ack_config_version(
    ctx: &ReducerContext,
    server_id: String,
    namespace: String,
    key: String,
    version: u64,
) -> Result<(), String> {
    validate_runtime_config_ack_identity(&server_id, &namespace, &key)?;
    let config = ctx
        .db
        .runtime_config()
        .iter()
        .find(|row| row.server_id == server_id && row.namespace == namespace && row.key == key);
    let Some(config) = config else {
        return Err("unknown config".to_string());
    };
    if config.version != version {
        return Err("config version mismatch".to_string());
    }

    let id = runtime_config_ack_id(&server_id, &namespace, &key);
    let ack = RuntimeConfigAck {
        id: id.clone(),
        server_id,
        namespace,
        key,
        version,
        acknowledged_at: ctx.timestamp,
    };
    if ctx.db.runtime_config_acks().id().find(id).is_some() {
        ctx.db.runtime_config_acks().id().update(ack);
    } else {
        ctx.db.runtime_config_acks().insert(ack);
    }
    Ok(())
}

fn validate_runtime_config_ack_identity(server_id: &str, namespace: &str, key: &str) -> Result<(), String> {
    if server_id.trim().is_empty() {
        return Err("runtime config ack server id is required".to_string());
    }
    if namespace.trim().is_empty() {
        return Err("runtime config ack namespace is required".to_string());
    }
    if key.trim().is_empty() {
        return Err("runtime config ack key is required".to_string());
    }
    Ok(())
}
